cwl_utils/sandboxjs.py

```diff
@@ -317,17 +319,21 @@ def new_js_proc(
                             [container_engine, "images", "-q", nodeimg],
                             universal_newlines=True,
                         )
+                    elif container_engine == "singularity":
+                        try:
+                            singularityimgs = glob.glob(os.environ['CWL_SINGULARITY_CACHE'] + '/node_slim.sif')
+                        except KeyError:
+                            singularityimgs = glob.glob(os.getcwd() + '/node_slim.sif')
                     elif container_engine != "singularity":
                         raise Exception(
                             f"Unknown container_engine: {container_engine}."
                         )
                     # if output is an empty string
                     if (
-                        container_engine == "singularity"
+                        len(singularityimgs) == 0
```

Comment:
you could write: `not singularityimgs` instead of testing whether the length is empty.

To remove the earlier `singularityimgs = [0]` and `dockerimgs="\n"` statements, you could change this if statement to:
```
need_singularity = container_engine == "singularity" and not singularityimgs
need_docker = container_engine != "singularity" and (len(dockerimgs.split("\n")) <= 1
if(
  need_singularity
  or need_docker
  or force_docker_pull)
):
```
I think this is more descriptive in what this if statement does. The need_singularity and need_docker statements can then also be set in the engine-testing branches above, so that they become simpler and flow becomes even more clear.